ci/create_database.py

```diff
@@ -0,0 +1,208 @@
+    print(f'creating secret {secret_name}')
+    await check_shell(
+        f'''
+kubectl -n {shq(namespace)} delete secret {shq(secret_name)}
+kubectl -n {shq(namespace)} create secret generic {shq(secret_name)} --from-file=sql-config.json --from-file=sql-config.cnf
+''')
+
+
+async def create_database(create_database_config):
+    with open('/sql-config/sql-config.json', 'r') as f:
+        sql_config = json.loads(f.read())
+
+    namespace = create_database_config['namespace']
+    scope = create_database_config['scope']
+    code_short_str = create_database_config['code_short_str']
+    database_name = create_database_config['database_name']
+    cant_create_database = create_database_config['cant_create_database']
+
+    if cant_create_database:
+        assert sql_config.get('db') is not None
+
+        await write_user_config(namespace, database_name, 'admin', sql_config)
+        await write_user_config(namespace, database_name, 'user', sql_config)
+        return
+
+    db = Database()
+    await db.async_init()
+
+    if scope == 'deploy':
+        # create if not exists
+        rows = db.execute_and_fetchall(
+            "SHOW DATABASES LIKE '{database_name}';")
+        rows = [row async for row in rows]
+        if len(rows) > 0:
+            assert len(rows) == 1
+            return
+
+        _name = database_name
+        admin_username = f'{database_name}-admin'
+        user_username = f'{database_name}-user'
+    else:
+        _name = f'{code_short_str}-{database_name}-{generate_token()}'
+        admin_username = generate_token()
+        user_username = generate_token()
+
+    admin_password = secrets.token_urlsafe(16)
+    user_password = secrets.token_urlsafe(16)
+
+    await db.just_execute(f'''
+CREATE DATABASE `{_name}`;
+
+CREATE USER '{admin_username}'@'%' IDENTIFIED BY '{admin_password}';
+GRANT ALL ON `{_name}`.* TO '{admin_username}'@'%';
+
+CREATE USER '{user_username}'@'%' IDENTIFIED BY '{user_password}';
+GRANT SELECT, INSERT, UPDATE, DELETE, EXECUTE ON `{_name}`.* TO '{user_username}'@'%';
+''')
+
+    await write_user_config(namespace, database_name, 'admin', {
+        'host': sql_config['host'],
+        'port': sql_config['port'],
+        'instance': sql_config['instance'],
+        'connection_name': sql_config['connection_name'],
+        'user': admin_username,
+        'password': admin_password,
+        'db': _name
+    })
+
+    await write_user_config(namespace, database_name, 'user', {
+        'host': sql_config['host'],
+        'port': sql_config['port'],
+        'instance': sql_config['instance'],
+        'connection_name': sql_config['connection_name'],
+        'user': user_username,
+        'password': user_password,
+        'db': _name
+    })
+
+
+async def migrate(database_name, db, i, migration):
+    print(f'applying migration {i} {migration}')
+
+    # version to migrate to
+    # the 0th migration migrates from 1 to 2
+    version = i + 2
```

Comment:
I don't understand why it's i + 2.

Reply:
I tried to explain it with the comment:

 > # the 0th migration migrates from 1 to 2

The first (empty) database version is 1.  `version` is the version to migration to.  The first migration step (at index i = 0) goes from version 1 to version 2.  So version = i + 2.  Make sense?

Reply:
Yes. I was being stupid. I thought it was causing version to skip from 0, 2, 4, 6, 8 etc. rather than 2, 3, 4, 5,... I should relook at my discrete math homework for the definition of even numbers.